source/data-modeling/field-behaviors.txt

```diff
@@ -0,0 +1,292 @@
+
+=========================
+Customize Field Behaviors
+=========================
+
+.. facet::
+   :name: genre
+   :values: reference
+
+.. meta::
+   :keywords: customize, attributes, optimize, model, configure, code example
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 1
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn how to customize the behavior of fields in {+odm+} models.
+
+Specify Default Values
+----------------------
+
+You can configure fields to have default values by using the ``default`` option.
+Default field values can be either fixed or ``Proc`` values.
+
+The following example specifies a fixed default value for the ``fulfill_by``
+field:
+
+.. literalinclude:: /includes/data-modeling/field-behaviors.rb
+   :language: ruby
+   :start-after: # start-field-default
+   :end-before: # end-field-default
+
+The following example specifies a ``Proc`` default value for the ``fulfill_by``
+field:
+
+.. literalinclude:: /includes/data-modeling/field-behaviors.rb
+   :language: ruby
+   :start-after: # start-field-default-processed
+   :end-before: # end-field-default-processed
+
+.. note::
+   
+   The driver evaluates default values that are not ``Proc`` instances when the
+   class *loads*. The driver evaluates ``Proc`` values when the document is
+   *instantiated*. The following default field values do not produce equivalent outcomes:
+
+   .. code-block:: ruby
+
+      # Time.now is set to the time the class is loaded
+      field :submitted_at, type: Time, default: Time.now
+
+      # Time.now is set to the time the document is instantiated
+      field :submitted_at, type: Time, default: ->{ Time.now }
+
+You can set a default value that depends on the document's state by using the
+``self`` keyword in a ``Proc`` instance. The following example sets the
+``fulfill_by`` default value to depend on the state of the ``submitted_at``
+field:
+
+.. literalinclude:: /includes/data-modeling/field-behaviors.rb
+   :language: ruby
+   :start-after: # start-field-default-self
+   :end-before: # end-field-default-self
+
+By default, {+odm+} applies ``Proc`` default values after setting and
+initializing all other attributes. To apply the default before setting the other
+attributes, set the ``pre_processed`` option to ``true``, as shown in the
+following example:
+
+.. literalinclude:: /includes/data-modeling/field-behaviors.rb
+   :language: ruby
+   :start-after: # start-field-default-pre-processed
+   :end-before: # end-field-default-pre-processed
+
+.. tip::
+
+   Always set the ``pre-processed`` option to ``true`` to set a
+   default ``Proc`` value for the ``_id`` field.
+
+Specify Storage Names
+---------------------
+
+You can specify a separate field name to store in the database, while still
+referring to the field by its original name in your application. This can be
+useful to save storage space, because MongoDB stores all field information along
```

Comment:
[s] Can potentially edit this for brevity and avoiding extra words:
```suggestion
referring to the field by its original name in your application. This can save storage space, because MongoDB stores all field information along
```